### PR TITLE
[G7T5-40] Skill components in DB

### PR DESCRIPTION
**Pull request by branfats**
# Description
Created skill for db, models and schemas. skill_id is a foreign key for role schema

Fixes # (issue)

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

Not Applicable

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: backend/app/models/__init__.py [hunk 1]
```diff
@@ -2,3 +2,4 @@
 from .registration import Registration
 from .role import Role
 from .staff import Staff
+from .skill import Skill
```
Comment (jonathantan1425) [inline backend/app/models/__init__.py:5]: Need EOF blank line (i.e. one empty line after) -- install the pre-commit hooks, it should automatically check for you

File: backend/app/models/role.py [hunk 2]
```diff
@@ -6,3 +6,4 @@
 class Role(Base):
     role_id = Column(Integer, primary_key=True, index=True)
     role_name = Column(String, nullable=False)
+    linked_skill = Column(Integer, ForeignKey("skill.skill_id"))
```
Comment (jonathantan1425) [inline backend/app/models/role.py:9]: Same EOF issue.
This role is the access control role for LMS, not the job "role" that we have. This linked skill shouldn't be here

File: backend/app/models/skill.py [hunk 3]
```diff
@@ -0,0 +1,8 @@
+from sqlalchemy import Column, Integer, String
+
+from app.db.base_class import Base
+
+
+class Skill(Base):
+    skill_id = Column(Integer, primary_key=True, index=True)
+    skill_name = Column(String, nullable=False)
```
Comment (jonathantan1425) [inline backend/app/models/skill.py:8]: Can confirm if these fields are standardised against the fields in the DB schema?
https://lucid.app/lucidchart/3927e942-08c3-44fc-b056-e4117c97d0a0/edit?beaconFlowId=36667FEA57D405BF&invitationId=inv_b24547aa-e507-4177-9804-29fe5da262b4&page=0_0#

**Review by jonathantan1425 — CHANGES_REQUESTED** [review]
Just double-check if the fields added are aligned with the DB schema, otherwise we'll have mismatched schema diagram and actual fields. Also missing the alembic version for the MySQL changes.

File: backend/app/models/__init__.py (hunk 1 above)
Comment (quinncheong) [reply backend/app/models/__init__.py]: Why do we need this ah? Is this for just style guide or there's semantic value

Comment (branfats) [reply backend/app/models/__init__.py]: I searched up why too and just did it for all files

https://stackoverflow.com/questions/2287967/why-is-it-recommended-to-have-empty-line-in-the-end-of-a-source-file

File: backend/app/models/role.py (hunk 2 above)
Comment (charmjj) [reply backend/app/models/role.py]: We need to create a new table for job role, this "role" refers to Staff/HR/Manager.

File: backend/app/models/__init__.py (hunk 1 above)
Comment (jonathantan1425) [reply backend/app/models/__init__.py]: ^ pretty much summarizes why, but its an in-built styling pattern applied in Python Black

File: backend/alembic/versions/0002_create_skills.py [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+Revision ID: 0002
+Revises: 0001
+Create Date: 2022-09-18 00:00:00
+
+"""
+from alembic import op
+
+# revision identifiers, used by Alembic.
+revision = "0002"
+down_revision = None
```
Comment (jonathantan1425) [inline backend/alembic/versions/0002_create_skills.py:12]: down revision should be 0001

File: backend/alembic/versions/0002_create_skills.py [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+branch_labels = None
+depends_on = None
+
+
+UPGRADE_SQL = """
+CREATE TABLE IF NOT EXISTS `role` (
+    `Role_ID` int NOT NULL AUTO_INCREMENT,
+    `Role_Name` varchar(20) NOT NULL,
+    PRIMARY KEY (`Role_ID`)
+) DEFAULT CHARSET=utf8;
```
Comment (jonathantan1425) [inline backend/alembic/versions/0002_create_skills.py:22]: These tables seem to conflict with 0001 (i.e. creating tables of the same name). Would be better to either alter the existing tables or drop them then recreate

File: backend/app/models/skill.py [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+
+from app.db.base_class import Base
+
+
+class Skill(Base):
+    skill_id = Column(Integer, primary_key=True, index=True)
+    course_id = Column(String, ForeignKey("course.course_id"), nullable=False) 
+    skill_name = Column(String, nullable=False)
+    skill_desc = Column(String)
+    Is_Active = Column(Boolean)
```
Comment (jonathantan1425) [inline backend/app/models/skill.py:11]: `Is_Active` variable shouldn't be capitalized

File: backend/app/schemas/role.py [hunk 7]
```diff
@@ -7,6 +7,7 @@
 class RoleBase(BaseModel):
     role_id: Optional[int]
     role_name: str
+    linked_skill: int #Skill ID foreign key
```
Comment (jonathantan1425) [inline backend/app/schemas/role.py:10]: Does this tally with alembic version 0002?

File: backend/app/models/skill.py (hunk 6 above)
Comment (branfats) [reply backend/app/models/skill.py:11]: ok will make changes to all ur comments accordingly

File: backend/app/schemas/role.py [hunk 8]
```diff
@@ -9,6 +9,7 @@ class RoleBase(BaseModel):
     role_name: str
 
 
+
```
Comment (quinncheong) [inline backend/app/schemas/role.py:12]: /nit Is this intended? Does pre-commit remove this @jonathantan1425 ?

Comment (jonathantan1425) [inline backend/app/schemas/role.py:12]: Is this extra line needed?

**Review by jonathantan1425 — APPROVED** [review]

**Review by quinncheong — APPROVED** [review]

File: backend/app/schemas/role.py (hunk 8 above)
Comment (jonathantan1425) [reply backend/app/schemas/role.py:12]: I think it should? I'm not very sure

Comment (branfats) [reply backend/app/schemas/role.py:12]: Not sure, this file wasn't done by me

**Pull request merged**